crates/noirc_frontend/src/monomorphisation/mod.rs

```diff
@@ -0,0 +1,605 @@
+        ast::Expression::Block(vecmap(statement_ids, |id| self.statement(id)))
+    }
+
+    fn unpack_pattern(
+        &mut self,
+        pattern: HirPattern,
+        value: ast::Expression,
+        typ: &HirType,
+    ) -> ast::Expression {
+        match pattern {
+            HirPattern::Identifier(ident) => {
+                let new_id = self.next_definition_id();
+                self.define_local(ident.id, new_id);
+                ast::Expression::Let(ast::Let {
+                    id: new_id,
+                    name: self.interner.definition_name(ident.id).to_owned(),
+                    expression: Box::new(value),
+                })
+            }
+            HirPattern::Mutable(pattern, _) => self.unpack_pattern(*pattern, value, typ),
+            HirPattern::Tuple(patterns, _) => {
+                let fields = unwrap_tuple_type(typ);
+                self.unpack_tuple_pattern(value, patterns.into_iter().zip(fields))
+            }
+            HirPattern::Struct(_, patterns, _) => {
+                let fields = unwrap_struct_type(typ);
+                let patterns = patterns.into_iter().map(|(ident, pattern)| {
+                    let typ = fields[&ident.0.contents].clone();
+                    (pattern, typ)
+                });
+                self.unpack_tuple_pattern(value, patterns)
+            }
+        }
+    }
+
+    fn unpack_tuple_pattern(
+        &mut self,
+        value: ast::Expression,
+        fields: impl Iterator<Item = (HirPattern, HirType)>,
+    ) -> ast::Expression {
+        let fresh_id = self.next_definition_id();
+
+        let mut definitions = vec![ast::Expression::Let(ast::Let {
+            id: fresh_id,
+            name: "_".into(),
+            expression: Box::new(value),
+        })];
+
+        for (i, (field_pattern, field_type)) in fields.into_iter().enumerate() {
+            let typ = Self::convert_type(&field_type);
+            let name = i.to_string();
+            let new_rhs =
+                ast::Expression::Ident(ast::Ident { location: None, id: fresh_id, name, typ });
+            let new_rhs = ast::Expression::ExtractTupleField(Box::new(new_rhs), i);
+            let new_expr = self.unpack_pattern(field_pattern, new_rhs, &field_type);
+            definitions.push(new_expr);
+        }
+
+        ast::Expression::Block(definitions)
+    }
+
+    /// A local (ie non-global) ident only
+    fn local_ident(&mut self, ident: HirIdent) -> ast::Ident {
+        let id = self.lookup_local(ident.id).unwrap();
+        let name = self.interner.definition_name(ident.id).to_owned();
+        let typ = Self::convert_type(&self.interner.id_type(ident.id));
+        ast::Ident { location: Some(ident.location), id, name, typ }
+    }
+
+    fn ident(&mut self, ident: HirIdent) -> ast::Expression {
+        match self.lookup_local(ident.id) {
+            Some(id) => {
+                // Normal local variable
```

Comment:
I think you should use local_ident() in that case instead of duplicating the code